File: workspace_interpreter/expr_lexer.py
"""Tokenizer for the expression language."""

from __future__ import annotations
from enum import Enum, auto
from dataclasses import dataclass


class TokenKind(Enum):
    # Literals
    IDENT = auto()
    NUMBER = auto()
    STRING = auto()
    COLOR = auto()
    # Keywords
    TRUE = auto()
    FALSE = auto()
    NULL = auto()
    NOT = auto()
    AND = auto()
    OR = auto()
    IN = auto()
    FUN = auto()
    LET = auto()
    IF = auto()
    THEN = auto()
    ELSE = auto()
    # Operators
    EQ = auto()        # ==
    NEQ = auto()       # !=
    LT = auto()        # <
    GT = auto()        # >
    LTE = auto()       # <=
    GTE = auto()       # >=
    QUESTION = auto()  # ?
    COLON = auto()     # :
    DOT = auto()       # .
    COMMA = auto()     # ,
    LPAREN = auto()    # (
    RPAREN = auto()    # )
    LBRACKET = auto()  # [
    RBRACKET = auto()  # ]
    PLUS = auto()      # +
    MINUS = auto()     # -
    STAR = auto()      # *
    SLASH = auto()     # /
    ARROW = auto()     # ->
    LARROW = auto()    # <-
    SEMICOLON = auto() # ;
    EQUALS = auto()    # =
    # Control
    EOF = auto()
    ERROR = auto()


_KEYWORDS = {
    "true": TokenKind.TRUE,
    "false": TokenKind.FALSE,
    "null": TokenKind.NULL,
    "not": TokenKind.NOT,
    "and": TokenKind.AND,
    "or": TokenKind.OR,
    "in": TokenKind.IN,
    "fun": TokenKind.FUN,
    "let": TokenKind.LET,
    "if": TokenKind.IF,
    "then": TokenKind.THEN,
    "else": TokenKind.ELSE,
}


@dataclass(slots=True)
class Token:
    kind: TokenKind
    value: str | float | None = None


def tokenize(source: str) -> list[Token]:
    """Tokenize an expression string into a list of Tokens."""
    tokens: list[Token] = []
    i = 0
    n = len(source)

    while i < n:
        c = source[i]

        # Whitespace
        if c in " \t":
            i += 1
            continue

        # Color literal: #rrggbb or #rgb
        if c == "#":
            j = i + 1
            while j < n and source[j] in "0123456789abcdefABCDEF":
                j += 1
            hex_len = j - i - 1
            if hex_len in (3, 6):
                tokens.append(Token(TokenKind.COLOR, source[i:j].lower()))
                i = j
                continue
            tokens.append(Token(TokenKind.ERROR, source[i:j]))
            i = j
            continue

        # String literal (double or single quotes)
        if c == '"' or c == "'":
            quote = c
            j = i + 1
            parts = []
            while j < n and source[j] != quote:
                if source[j] == "\\" and j + 1 < n:
                    parts.append(source[j + 1])
                    j += 2
                else:
                    parts.append(source[j])
                    j += 1
            if j < n:
                j += 1  # consume closing quote
            tokens.append(Token(TokenKind.STRING, "".join(parts)))
            i = j
            continue

        # Number (digits only — unary minus is handled as an operator)
        if c.isdigit():
            j = i
            while j < n and source[j].isdigit():
                j += 1
            if j < n and source[j] == ".":
                j += 1
                while j < n and source[j].isdigit():
                    j += 1
                tokens.append(Token(TokenKind.NUMBER, float(source[i:j])))
            else:
                tokens.append(Token(TokenKind.NUMBER, int(source[i:j])))
            i = j
            continue

        # Identifier / keyword
        if c.isalpha() or c == "_":
            j = i + 1
            while j < n and (source[j].isalnum() or source[j] == "_"):
                j += 1
            word = source[i:j]
            kind = _KEYWORDS.get(word, TokenKind.IDENT)
            tokens.append(Token(kind, word))
            i = j
            continue

        # Multi-character operators (order matters)
        if c == "=" and i + 1 < n and source[i + 1] == "=":
            tokens.append(Token(TokenKind.EQ))
            i += 2
            continue
        if c == "!" and i + 1 < n and source[i + 1] == "=":
            tokens.append(Token(TokenKind.NEQ))
            i += 2
            continue
        # <- : no space between < and -
        if c == "<" and i + 1 < n and source[i + 1] == "-":
            tokens.append(Token(TokenKind.LARROW))
            i += 2
            continue
        if c == "<" and i + 1 < n and source[i + 1] == "=":
            tokens.append(Token(TokenKind.LTE))
            i += 2
            continue
        if c == ">" and i + 1 < n and source[i + 1] == "=":
            tokens.append(Token(TokenKind.GTE))
            i += 2
            continue
        # -> : no space between - and >
        if c == "-" and i + 1 < n and source[i + 1] == ">":
            tokens.append(Token(TokenKind.ARROW))
            i += 2
            continue

        # Single-character operators
        if c == "<":
            tokens.append(Token(TokenKind.LT))
        elif c == ">":
            tokens.append(Token(TokenKind.GT))
        elif c == "=":
            tokens.append(Token(TokenKind.EQUALS))
        elif c == "?":
            tokens.append(Token(TokenKind.QUESTION))
        elif c == ":":
            tokens.append(Token(TokenKind.COLON))
        elif c == ".":
            tokens.append(Token(TokenKind.DOT))
        elif c == ",":
            tokens.append(Token(TokenKind.COMMA))
        elif c == ";":
            tokens.append(Token(TokenKind.SEMICOLON))
        elif c == "(":
            tokens.append(Token(TokenKind.LPAREN))
        elif c == ")":
            tokens.append(Token(TokenKind.RPAREN))
        elif c == "[":
            tokens.append(Token(TokenKind.LBRACKET))
        elif c == "]":
            tokens.append(Token(TokenKind.RBRACKET))
        elif c == "+":
            tokens.append(Token(TokenKind.PLUS))
        elif c == "-":
            tokens.append(Token(TokenKind.MINUS))
        elif c == "*":
            tokens.append(Token(TokenKind.STAR))
        elif c == "/":
            tokens.append(Token(TokenKind.SLASH))
        else:
            tokens.append(Token(TokenKind.ERROR, c))
        i += 1

    tokens.append(Token(TokenKind.EOF))
    return tokens

File: jas_dioxus/src/interpreter/expr_lexer.rs
//! Tokenizer for the expression language.

#[derive(Debug, Clone, PartialEq)]
pub enum TokenKind {
    Ident(String),
    Number(f64),
    Str(String),
    Color(String),
    True,
    False,
    Null,
    Not,
    And,
    Or,
    In,
    Fun,
    Let,
    If,
    Then,
    Else,
    Eq,        // ==
    Neq,       // !=
    Lt,        // <
    Gt,        // >
    Lte,       // <=
    Gte,       // >=
    Question,
    Colon,
    Dot,
    Comma,
    LParen,
    RParen,
    LBracket,
    RBracket,
    Plus,      // +
    Minus,     // -
    Star,      // *
    Slash,     // /
    Arrow,     // ->
    LArrow,    // <-
    Semicolon, // ;
    Equals,    // =
    Eof,
    Error(String),
}

#[derive(Debug, Clone)]
pub struct Token {
    pub kind: TokenKind,
}

pub fn tokenize(source: &str) -> Vec<Token> {
    let mut tokens = Vec::new();
    let chars: Vec<char> = source.chars().collect();
    let n = chars.len();
    let mut i = 0;

    while i < n {
        let c = chars[i];

        // Whitespace
        if c == ' ' || c == '\t' {
            i += 1;
            continue;
        }

        // Color literal: #rrggbb or #rgb
        if c == '#' {
            let mut j = i + 1;
            while j < n && chars[j].is_ascii_hexdigit() {
                j += 1;
            }
            let hex_len = j - i - 1;
            if hex_len == 3 || hex_len == 6 {
                let s: String = chars[i..j].iter().collect();
                tokens.push(Token { kind: TokenKind::Color(s.to_lowercase()) });
                i = j;
                continue;
            }
            let s: String = chars[i..j].iter().collect();
            tokens.push(Token { kind: TokenKind::Error(s) });
            i = j;
            continue;
        }

        // String literal
        if c == '"' {
            let mut j = i + 1;
            let mut parts = String::new();
            while j < n && chars[j] != '"' {
                if chars[j] == '\\' && j + 1 < n {
                    parts.push(chars[j + 1]);
                    j += 2;
                } else {
                    parts.push(chars[j]);
                    j += 1;
                }
            }
            if j < n { j += 1; } // consume closing "
            tokens.push(Token { kind: TokenKind::Str(parts) });
            i = j;
            continue;
        }

        // Number (digits only — unary minus handled as operator)
        if c.is_ascii_digit() {
            let start = i;
            while i < n && chars[i].is_ascii_digit() { i += 1; }
            if i < n && chars[i] == '.' {
                i += 1;
                while i < n && chars[i].is_ascii_digit() { i += 1; }
                let s: String = chars[start..i].iter().collect();
                tokens.push(Token { kind: TokenKind::Number(s.parse().unwrap_or(0.0)) });
            } else {
                let s: String = chars[start..i].iter().collect();
                tokens.push(Token { kind: TokenKind::Number(s.parse().unwrap_or(0.0)) });
            }
            continue;
        }

        // Identifier / keyword
        if c.is_ascii_alphabetic() || c == '_' {
            let start = i;
            while i < n && (chars[i].is_ascii_alphanumeric() || chars[i] == '_') { i += 1; }
            let word: String = chars[start..i].iter().collect();
            let kind = match word.as_str() {
                "true" => TokenKind::True,
                "false" => TokenKind::False,
                "null" => TokenKind::Null,
                "not" => TokenKind::Not,
                "and" => TokenKind::And,
                "or" => TokenKind::Or,
                "in" => TokenKind::In,
                "fun" => TokenKind::Fun,
                "let" => TokenKind::Let,
                "if" => TokenKind::If,
                "then" => TokenKind::Then,
                "else" => TokenKind::Else,
                _ => TokenKind::Ident(word),
            };
            tokens.push(Token { kind });
            continue;
        }

        // Two-character operators (order matters for greedy matching)
        if i + 1 < n {
            let next = chars[i + 1];
            match (c, next) {
                ('=', '=') => { tokens.push(Token { kind: TokenKind::Eq }); i += 2; continue; }
                ('!', '=') => { tokens.push(Token { kind: TokenKind::Neq }); i += 2; continue; }
                // <- must come before < alone (greedy)
                ('<', '-') => { tokens.push(Token { kind: TokenKind::LArrow }); i += 2; continue; }
                ('<', '=') => { tokens.push(Token { kind: TokenKind::Lte }); i += 2; continue; }
                ('>', '=') => { tokens.push(Token { kind: TokenKind::Gte }); i += 2; continue; }
                // -> must come before - alone (greedy)
                ('-', '>') => { tokens.push(Token { kind: TokenKind::Arrow }); i += 2; continue; }
                _ => {}
            }
        }

        // Single-character operators
        match c {
            '<' => tokens.push(Token { kind: TokenKind::Lt }),
            '>' => tokens.push(Token { kind: TokenKind::Gt }),
            '=' => tokens.push(Token { kind: TokenKind::Equals }),
            '?' => tokens.push(Token { kind: TokenKind::Question }),
            ':' => tokens.push(Token { kind: TokenKind::Colon }),
            '.' => tokens.push(Token { kind: TokenKind::Dot }),
            ',' => tokens.push(Token { kind: TokenKind::Comma }),
            ';' => tokens.push(Token { kind: TokenKind::Semicolon }),
            '(' => tokens.push(Token { kind: TokenKind::LParen }),
            ')' => tokens.push(Token { kind: TokenKind::RParen }),
            '[' => tokens.push(Token { kind: TokenKind::LBracket }),
            ']' => tokens.push(Token { kind: TokenKind::RBracket }),
            '+' => tokens.push(Token { kind: TokenKind::Plus }),
            '-' => tokens.push(Token { kind: TokenKind::Minus }),
            '*' => tokens.push(Token { kind: TokenKind::Star }),
            '/' => tokens.push(Token { kind: TokenKind::Slash }),
            _ => tokens.push(Token { kind: TokenKind::Error(c.to_string()) }),
        }
        i += 1;
    }

    tokens.push(Token { kind: TokenKind::Eof });
    tokens
}
